File: core/src/main/java/com/zp4rker/iab/prompts/ExplorerSetupPrompt.java
package com.zp4rker.iab.prompts;

import com.zp4rker.iab.api.Explorer;
import org.bukkit.conversations.ConversationContext;
import org.bukkit.conversations.MessagePrompt;
import org.bukkit.conversations.Prompt;
import org.bukkit.conversations.ValidatingPrompt;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ExplorerSetupPrompt extends ValidatingPrompt {
    @NotNull
    @Override
    public String getPromptText(@NotNull ConversationContext conversationContext) {
        return "Please enter your explorer name";
    }

    @Nullable
    @Override
    protected Prompt acceptValidatedInput(@NotNull ConversationContext conversationContext, @NotNull String s) {
        return new ConfirmationPrompt(s);
    }

    @Nullable
    @Override
    protected String getFailedValidationText(@NotNull ConversationContext context, @NotNull String invalidInput) {
        return "Explorer names have to be 35 characters or less";
    }

    @Override
    protected boolean isInputValid(@NotNull ConversationContext conversationContext, @NotNull String s) {
        return !s.isEmpty() && s.length() <= 35;
    }

    private static class ConfirmationPrompt extends MessagePrompt {
        private final String explorerName;

        public ConfirmationPrompt(String explorerName) {
            this.explorerName = explorerName;
        }

        @Nullable
        @Override
        protected Prompt getNextPrompt(@NotNull ConversationContext conversationContext) {
            return Prompt.END_OF_CONVERSATION;
        }

        @NotNull
        @Override
        public String getPromptText(@NotNull ConversationContext conversationContext) {
            Explorer explorer = new Explorer((Player) conversationContext.getForWhom(), explorerName);

            if (!explorer.save()) {
                return "Failed to set your explorer name! Please try again";
            }

            return "Set your explorer name to " + explorerName;
        }
    }

    @Override
    public boolean blocksForInput(@NotNull ConversationContext context) {
        return true;
    }
}

File: core/src/main/kotlin/com/zp4rker/iab/prompts/ExplorerSetupPrompt.kt
package com.zp4rker.iab.prompts

import com.zp4rker.iab.api.Explorer
import com.zp4rker.iab.utils.Lang
import com.zp4rker.iab.utils.legacy
import net.kyori.adventure.text.TextReplacementConfig
import org.bukkit.conversations.ConversationContext
import org.bukkit.conversations.Prompt
import org.bukkit.conversations.ValidatingPrompt
import org.bukkit.entity.Player

class ExplorerSetupPrompt : ValidatingPrompt() {
    var explorer: Explorer? = null

    override fun getPromptText(ctx: ConversationContext): String {
        return Lang.getMessage("explorer-setup.prompt").legacy()
    }

    override fun getFailedValidationText(context: ConversationContext, invalidInput: String): String {
        return Lang.getMessage("explorer-setup.error").legacy()
    }

    override fun isInputValid(ctx: ConversationContext, input: String) = input.length <= 35
    override fun acceptValidatedInput(ctx: ConversationContext, input: String): Prompt {
        explorer = Explorer(ctx.forWhom as Player, input)

        return if (explorer?.save() == true) {
            var message = Lang.getMessage("explorer-setup.success")
            val tr = TextReplacementConfig.builder().matchLiteral("%name%").replacement(input).build()
            message = message.replaceText(tr)
            SingleMessagePrompt(message.legacy())
        } else {
            SingleMessagePrompt(Lang.getMessage("explorer-setup.fail").legacy())
        }
    }

    override fun blocksForInput(ctx: ConversationContext) = true
}